File: applications/IgaApplication/tests/cpp_tests/test_support_solid_IGA_condition.cpp
//    |  /           |
//    ' /   __| _` | __|  _ \   __|
//    . \  |   (   | |   (   |\__ `
//   _|\_\_|  \__,_|\__|\___/ ____/
//                   Multi-Physics
//
//  License:         BSD License
//                   Kratos default license: kratos/license.txt
//
//  Main authors:    Andrea Gorgi
//

// Project includes
#include "containers/model.h"
#include "testing/testing.h"
#include "test_creation_utility.h"
#include "custom_conditions/support_solid_iga_condition.h"
#include "iga_structural_mechanics_fast_suite.h"

namespace Kratos::Testing
{
namespace
{
    typedef std::size_t SizeType;

    typename Condition::Pointer GetSupportSolidIGACondition(
        ModelPart& rModelPart, 
        SizeType PolynomialDegree, 
        IntegrationPoint<3> IntegrationPoint)
    {
        // Set the condition properties
        auto p_cond_prop = rModelPart.CreateNewProperties(0);
        p_cond_prop->SetValue(PENALTY_FACTOR, -1.0); // Penalty-free formulation
        p_cond_prop->SetValue(THICKNESS, 1.0);
        p_cond_prop->SetValue(YOUNG_MODULUS, 100);
        p_cond_prop->SetValue(POISSON_RATIO, 0.3);

        // Costitutive law
        const auto& r_clone_cl = KratosComponents<ConstitutiveLaw>::Get("LinearElasticPlaneStrain2DLaw");
        p_cond_prop->SetValue(CONSTITUTIVE_LAW, r_clone_cl.Clone());

        auto p_quadrature_point = TestCreationUtility::GetQuadraturePointGeometryOnCurve(
            rModelPart, PolynomialDegree, IntegrationPoint);

        return Kratos::make_intrusive<SupportSolidIGACondition>(1, p_quadrature_point, p_cond_prop);
    }
}

// Test per SupportSolidIGACondition con p=3
KRATOS_TEST_CASE_IN_SUITE(SupportSolidIGAConditionP3, KratosIgaSMFastSuite)
{
    Model model;
    auto& r_model_part = model.CreateModelPart("ModelPart");

    // Set buffer size
    r_model_part.SetBufferSize(2);

    // Aggiunta variabili
    r_model_part.AddNodalSolutionStepVariable(DISPLACEMENT);
    r_model_part.GetProcessInfo().SetValue(DOMAIN_SIZE, 2);

    const auto& r_process_info = r_model_part.GetProcessInfo();

    IntegrationPoint<3> integration_point(0.333333333333333, 0.05, 0.0, 0.086963711284364);
    auto p_support_condition = GetSupportSolidIGACondition(r_model_part, 3, integration_point);

    Vector u_D = ZeroVector(3);
    u_D[0] = 0.1; // Displacement in x-direction
    u_D[1] = -0.5; // Displacement in y-direction
    p_support_condition->SetValue(DISPLACEMENT, u_D);

    // Aggiunta DOFs
    for (auto& r_node : r_model_part.Nodes()) {
        r_node.AddDof(DISPLACEMENT_X);
        r_node.AddDof(DISPLACEMENT_Y);
    }

    p_support_condition->Initialize(r_process_info);
    p_support_condition->InitializeSolutionStep(r_process_info);
    p_support_condition->FinalizeSolutionStep(r_process_info);

    Matrix left_hand_side_matrix;
    Vector right_hand_side_vector;
    p_support_condition->CalculateLocalSystem(left_hand_side_matrix, right_hand_side_vector, r_process_info);

    // Check LHS e RHS
    const double tolerance = 1.0e-4;

    const std::vector<std::vector<double>> expected_LHS = { 
       {0, 0.596275, 0, 2.68324, 0, 2.23603, 0, 0.521741, 2.93641, 0.0313829, 4.40462, 0.141223, 2.20231, 0.117686, 0.367052, 0.0274601}, 
        {-0.596275, 0, 1.78883, 0, 2.23603, 0, 0.596275, 0, -0.0313829, 10.2774, 0.0941487, 15.4162, 0.117686, 7.70808, 0.0313829, 1.28468}, 
        {0, -1.78883, 0, -1.34055e-15, 0, 1.34162, 1.11022e-16, 0.447207, 4.40462, -0.0941487, 6.60693, -2.57848e-17, 3.30346, 0.0706116, 0.550577, 0.0235372}, 
        {-2.68324, 0, 1.34055e-15, 0, 2.01243, 1.77636e-15, 0.67081, 0, -0.141223, 15.4162, 1.15325e-16, 23.1243, 0.105917, 11.5621, 0.0353058, 1.92702}, 
        {0, -2.23603, 0, -2.01243, 0, -0.335405, 5.55112e-17, 0.0559008, 2.20231, -0.117686, 3.30346, -0.105917, 1.65173, -0.0176529, 0.275289, 0.00294215}, 
        {-2.23603, 0, -1.34162, -1.77636e-15, 0.335405, 0, 0.223603, 0, -0.117686, 7.70808, -0.0706116, 11.5621, 0.0176529, 5.78106, 0.0117686, 0.963511}, 
        {0, -0.596275, -1.11022e-16, -0.67081, -5.55112e-17, -0.223603, 0, -0.0186336, 0.367052, -0.0313829, 0.550577, -0.0353058, 0.275289, -0.0117686, 0.0458815, -0.000980716}, 
        {-0.521741, 0, -0.447207, 0, -0.0559008, 0, 0.0186336, 0, -0.0274601, 1.28468, -0.0235372, 1.92702, -0.00294215, 0.963511, 0.000980716, 0.160585}, 
        {-2.93641, 0.0313829, -4.40462, 0.141223, -2.20231, 0.117686, -0.367052, 0.0274601, 0, 0.00165173, 0, 0.0074328, 0, 0.006194, 0, 0.00144527}, 
        {-0.0313829, -10.2774, 0.0941487, -15.4162, 0.117686, -7.70808, 0.0313829, -1.28468, -0.00165173, 0, 0.0049552, 0, 0.006194, 0, 0.00165173, -1.38778e-17}, 
        {-4.40462, -0.0941487, -6.60693, -1.15325e-16, -3.30346, 0.0706116, -0.550577, 0.0235372, 0, -0.0049552, 0, -3.71343e-18, 0, 0.0037164, 0, 0.0012388}, 
        {-0.141223, -15.4162, 2.57848e-17, -23.1243, 0.105917, -11.5621, 0.0353058, -1.92702, -0.0074328, 0, 3.71343e-18, 0, 0.0055746, 1.11022e-16, 0.0018582, 0}, 
        {-2.20231, -0.117686, -3.30346, -0.105917, -1.65173, -0.0176529, -0.275289, 0.00294215, 0, -0.006194, 0, -0.0055746, 0, -0.000929099, 0, 0.00015485}, 
        {-0.117686, -7.70808, -0.0706116, -11.5621, 0.0176529, -5.78106, 0.0117686, -0.963511, -0.006194, 0, -0.0037164, -1.11022e-16, 0.000929099, 0, 0.0006194, -6.93889e-18}, 
        {-0.367052, -0.0313829, -0.550577, -0.0353058, -0.275289, -0.0117686, -0.0458815, -0.000980716, 0, -0.00165173, 0, -0.0018582, 0, -0.0006194, 0, -5.16166e-05}, 
        {-0.0274601, -1.28468, -0.0235372, -1.92702, -0.00294215, -0.963511, 0.000980716, -0.160585, -0.00144527, 1.38778e-17, -0.0012388, 0, -0.00015485, 6.93889e-18, 5.16166e-05, 0}
    };
    const std::vector<double> expected_RHS = {
        -2.18648,
        -16.9195,
        1.48656,
        -26.0148,
        3.12642,
        -13.3251,
        0.91826,
        -2.27382,
        -1.15828,
        17.3655,
        -1.48656,
        26.0148,
        -0.617851,
        12.9907,
        -0.0820705,
        2.16232
    };

    // std::cout << "LHS matrix:" << std::endl;
    // for (SizeType i = 0; i < left_hand_side_matrix.size1(); ++i) {
    //     std::cout << "{";
    //     for (SizeType j = 0; j < left_hand_side_matrix.size2(); ++j) {
    //         std::cout << left_hand_side_matrix(i,j);
    //         if (j < left_hand_side_matrix.size2() - 1) {
    //             std::cout << ", ";
    //         }
    //     }
    //     std::cout << "}";
    //     if (i < left_hand_side_matrix.size1() - 1) {
    //         std::cout << ", ";
    //     }
    //     std::cout << std::endl;
    // }

    // std::cout << "RHS vector:" << std::endl;
    // for (SizeType i = 0; i < right_hand_side_vector.size(); ++i) {
    //     std::cout << right_hand_side_vector(i) << "," << std::endl;
    // }
    // for (SizeType i = 0; i < left_hand_side_matrix.size1(); ++i) {
    //     for (SizeType j = 0; j < left_hand_side_matrix.size2(); ++j) {
    //         KRATOS_EXPECT_NEAR(left_hand_side_matrix(i,j), expected_LHS[i][j], tolerance);
    //     }
    // }

    for (SizeType i = 0; i < right_hand_side_vector.size(); ++i) {
        KRATOS_EXPECT_NEAR(right_hand_side_vector(i), expected_RHS[i], tolerance);
    }
}
}

File: applications/IgaApplication/tests/test_support_solid_IGA_condition.py
import KratosMultiphysics as KM
import KratosMultiphysics.IgaApplication as IGA
import KratosMultiphysics.StructuralMechanicsApplication as SMA
import KratosMultiphysics.KratosUnittest as KratosUnittest

from test_creation_utility import TestCreationUtility

class TestSupportSolidIGACondition(KratosUnittest.TestCase):

    def create_condition(self, model_part, degree, integration_point):
        # Define properties
        props = model_part.CreateNewProperties(0)
        props.SetValue(IGA.PENALTY_FACTOR, -1.0)  # penalty-free formulation
        props.SetValue(KM.THICKNESS, 1.0)
        props.SetValue(KM.YOUNG_MODULUS, 100)
        props.SetValue(KM.POISSON_RATIO, 0.3)

        # Assign constitutive law
        law = SMA.LinearElasticPlaneStrain2DLaw()
        props.SetValue(KM.CONSTITUTIVE_LAW, law)

        # Create quadrature point geometry on a curve
        geometry = TestCreationUtility.GetQuadraturePointGeometryOnCurveP2(model_part, integration_point)

        # Create the condition and add it to the model part
        condition = model_part.CreateNewCondition("SupportSolidCondition", 1, geometry, props)
        model_part.AddCondition(condition)

        return condition

    def test_SupportSolidIGAConditionP2(self):
        model = KM.Model()
        model_part = model.CreateModelPart("ModelPart")
        model_part.SetBufferSize(2)

        # Add required variables
        model_part.AddNodalSolutionStepVariable(KM.DISPLACEMENT)
        process_info = model_part.ProcessInfo

        # Define integration point in parametric space
        ipt = [0.333333333333333, 0.05, 0.0, 0.086963711284364]

        # Create the support condition
        condition = self.create_condition(model_part, 3, ipt)

        # Set prescribed displacement value
        u_D = KM.Vector(3)
        u_D[0] = 0.1
        u_D[1] = -0.5
        u_D[2] = 0.0
        condition.SetValue(KM.DISPLACEMENT, u_D)

        # Add degrees of freedom
        for node in model_part.Nodes:
            node.AddDof(KM.DISPLACEMENT_X)
            node.AddDof(KM.DISPLACEMENT_Y)

        # Initialize and step the condition
        condition.Initialize(process_info)
        
        # Compute local system
        lhs = KM.Matrix()
        rhs = KM.Vector()
        condition.CalculateLocalSystem(lhs, rhs, process_info)

        # Check the RHS
        expected_RHS = [
            -2.7362,-4.54032,1.79177,-5.14405,1.57993,-1.43695,-0.585333,4.72503,-0.108705,4.66148,0.0919808,1.14948,-0.0232275,0.261263,-0.0106846,0.25959,0.00046455,0.0644795
        ]
        tolerance = 1e-4

        self.assertEqual(rhs.Size(), len(expected_RHS))
        for i in range(rhs.Size()):
            self.assertAlmostEqual(rhs[i], expected_RHS[i], delta=tolerance)

        expected_LHS = [
            (0,0.807208,2.22045e-16,3.22883,1.66533e-16,1.41261,1.19255,0.0849692,1.19255,0.339877,0.298138,0.148696,0.0627658,0.00223603,0.0627658,0.00894413,0.0156915,0.00391306),(-0.807208,0,2.82523,8.88178e-16,1.61442,2.22045e-16,-0.0849692,4.17393,0.297392,4.17393,0.169938,1.04348,-0.00223603,0.21968,0.00782611,0.21968,0.00447207,0.0549201),(-2.22045e-16,-2.82523,0,-0.403604,5.55112e-17,0.504505,1.19255,-0.297392,1.19255,-0.0424846,0.298138,0.0531058,0.0627658,-0.00782611,0.0627658,-0.00111802,0.0156915,0.00139752),(-3.22883,-8.88178e-16,0.403604,0,1.00901,1.11022e-16,-0.339877,4.17393,0.0424846,4.17393,0.106212,1.04348,-0.00894413,0.21968,0.00111802,0.21968,0.00279504,0.0549201),(-1.66533e-16,-1.61442,-5.55112e-17,-1.00901,0,-0.100901,0.298138,-0.169938,0.298138,-0.106212,0.0745344,-0.0106212,0.0156915,-0.00447207,0.0156915,-0.00279504,0.00392286,-0.000279504),(-1.41261,-2.22045e-16,-0.504505,-1.11022e-16,0.100901,0,-0.148696,1.04348,-0.0531058,1.04348,0.0106212,0.26087,-0.00391306,0.0549201,-0.00139752,0.0549201,0.000279504,0.01373),(-1.19255,0.0849692,-1.19255,0.339877,-0.298138,0.148696,0,0.00894413,0,0.0357765,-3.46945e-18,0.0156522,0.00330346,0.000235372,0.00330346,0.000941487,0.000825866,0.000411901),(-0.0849692,-4.17393,0.297392,-4.17393,0.169938,-1.04348,-0.00894413,0,0.0313045,-5.55112e-17,0.0178883,0,-0.000235372,0.0115621,0.000823802,0.0115621,0.000470744,0.00289053),(-1.19255,-0.297392,-1.19255,-0.0424846,-0.298138,0.0531058,0,-0.0313045,0,-0.00447207,0,0.00559008,0.00330346,-0.000823802,0.00330346,-0.000117686,0.000825866,0.000147107),(-0.339877,-4.17393,0.0424846,-4.17393,0.106212,-1.04348,-0.0357765,5.55112e-17,0.00447207,0,0.0111802,0,-0.000941487,0.0115621,0.000117686,0.0115621,0.000294215,0.00289053),(-0.298138,-0.169938,-0.298138,-0.106212,-0.0745344,-0.0106212,3.46945e-18,-0.0178883,0,-0.0111802,0,-0.00111802,0.000825866,-0.000470744,0.000825866,-0.000294215,0.000206467,-2.94215e-05),(-0.148696,-1.04348,-0.0531058,-1.04348,0.0106212,-0.26087,-0.0156522,0,-0.00559008,0,0.00111802,0,-0.000411901,0.00289053,-0.000147107,0.00289053,2.94215e-05,0.000722633),(-0.0627658,0.00223603,-0.0627658,0.00894413,-0.0156915,0.00391306,-0.00330346,0.000235372,-0.00330346,0.000941487,-0.000825866,0.000411901,0,6.194e-06,2.71051e-20,2.4776e-05,0,1.08395e-05),(-0.00223603,-0.21968,0.00782611,-0.21968,0.00447207,-0.0549201,-0.000235372,-0.0115621,0.000823802,-0.0115621,0.000470744,-0.00289053,-6.194e-06,0,2.1679e-05,1.0842e-19,1.2388e-05,0),(-0.0627658,-0.00782611,-0.0627658,-0.00111802,-0.0156915,0.00139752,-0.00330346,-0.000823802,-0.00330346,-0.000117686,-0.000825866,0.000147107,-2.71051e-20,-2.1679e-05,0,-3.097e-06,-6.77626e-21,3.87125e-06),(-0.00894413,-0.21968,0.00111802,-0.21968,0.00279504,-0.0549201,-0.000941487,-0.0115621,0.000117686,-0.0115621,0.000294215,-0.00289053,-2.4776e-05,-1.0842e-19,3.097e-06,0,7.7425e-06,-2.71051e-20),(-0.0156915,-0.00447207,-0.0156915,-0.00279504,-0.00392286,-0.000279504,-0.000825866,-0.000470744,-0.000825866,-0.000294215,-0.000206467,-2.94215e-05,0,-1.2388e-05,6.77626e-21,-7.7425e-06,0,-7.7425e-07),(-0.00391306,-0.0549201,-0.00139752,-0.0549201,0.000279504,-0.01373,-0.000411901,-0.00289053,-0.000147107,-0.00289053,2.94215e-05,-0.000722633,-1.08395e-05,0,-3.87125e-06,2.71051e-20,7.7425e-07,0)
        ]
        for i in range(lhs.Size1()):
            for j in range(lhs.Size2()):
                self.assertAlmostEqual(lhs[i, j], expected_LHS[i][j], delta=tolerance)

if __name__ == '__main__':
    KratosUnittest.main()